src/radio_client.c

```diff
@@ -0,0 +1,1121 @@
+    GHashTable* pending;        /* Requests in PENDING state  */
+    RadioRequest* queue_first;  /* First QUEUED request */
+    RadioRequest* queue_last;   /* Last QUEUED request */
+    RadioRequest* block_req;
+    RadioRequestGroup* owner;
+    GSList* owner_queue;
+
+    gint64 next_deadline;       /* When the first PENDING request expires */
+    guint default_timeout_ms;   /* Default PENDING timeout, milliseconds */
+    guint timeout_id;
+} RadioClientObject;
+
+typedef GObjectClass RadioClientObjectClass;
+G_DEFINE_TYPE(RadioClientObject, radio_client_object, G_TYPE_OBJECT)
+
+#define PARENT_CLASS radio_client_object_parent_class
+#define THIS_TYPE (radio_client_object_get_type())
+#define THIS(obj) G_TYPE_CHECK_INSTANCE_CAST(obj, THIS_TYPE, RadioClientObject)
+
+enum radio_client_signal {
+    SIGNAL_INDICATION,
+    SIGNAL_OWNER,
+    SIGNAL_DEATH,
+    SIGNAL_CONNECTED,
+    SIGNAL_COUNT
+};
+
+#define SIGNAL_INDICATION_NAME "radio-client-indication"
+#define SIGNAL_OWNER_NAME      "radio-client-owner"
+#define SIGNAL_DEATH_NAME      "radio-client-death"
+#define SIGNAL_CONNECTED_NAME  "radio-client-connected"
+
+static guint radio_client_signals[SIGNAL_COUNT] = { 0 };
+
+static inline RadioClientObject* radio_client_cast(RadioClient* client)
+    { return client ? G_CAST(client, RadioClientObject, pub) : NULL; }
+static inline gboolean radio_client_can_retry(RadioRequest* req)
+    { return req->max_retries < 0 || req->max_retries > req->retry_count; }
+
+static
+void
+radio_client_object_reset_timeout(
+    RadioClientObject* self);
+
+/*==========================================================================*
+ * Implementation
+ *==========================================================================*/
+
+static
+guint32
+radio_client_reserve_serial(
+    RadioClientObject* self)
+{
+    do { self->last_serial++; }
+    while (!self->last_serial ||
+        g_hash_table_contains(self->requests, KEY(self->last_serial)));
+    return self->last_serial;
+}
+
+static
+guint
+radio_client_object_timeout_ms(
+    RadioClientObject* self,
+    RadioRequest* req)
+{
+    return req->timeout_ms ? req->timeout_ms : self->default_timeout_ms;
+}
+
+static
+void
+radio_client_cancel_transaction(
+    RadioClientObject* self,
+    RadioRequest* req)
+{
+    if (req->tx_id) {
+        radio_instance_cancel_request(self->pub.instance, req->tx_id);
+        req->tx_id = 0;
+    }
+}
+
+static
+void
+radio_client_deactivate_request(
+    RadioClientObject* self,
+    RadioRequest* req)
+{
+    g_hash_table_remove(self->pending, KEY(req->serial2));
+    g_hash_table_remove(self->pending, KEY(req->serial));
+    g_hash_table_remove(self->active, KEY(req->serial2));
+    g_hash_table_remove(self->active, KEY(req->serial));
+    if (req->state == RADIO_REQUEST_STATE_QUEUED) {
+        RadioRequest* ptr = self->queue_first;
+        RadioRequest* prev = NULL;
+
+        /* Remove it from the queue */
+        while (ptr) {
```

Comment:
I would use a for loop, but purely a matter of taste.

Reply:
(shame that GQueue doesn't have a _remove_link, otherwise it could've been used with some klugery to avoid writing custom list handling).

Reply:
One of the motivations for custom list handling was to avoid allocating extra memory blocks for links. Although I have to admit that it does smell like over-optimization.

Reply:
Yeah, I get that, and that was what the klugery meant; you could have:
```
struct client {
    GQueue requests;
}
```

and
```
struct request {
    GList listitem;
}
```

and then do:
```
g_queue_push_tail_link(&client->requests, &request->listitem);
```

Sprinkle a little pointer arithmetic on top.

Hmm, actually GQueue does have g_queue_remove_link, it's just called g_queue_unlink().

Oh well, up to you, the custom list comes with type safety, so it's not that bad.

What could be optimized, though, is to skip the O(n) lookup and instead do O(1) unlinking, like:
```
if (req->queue_next)
    req->queue_next->queue_prev = req->queue_prev;
else
    client->queue_last = req->queue_prev;
```
and similarily for prev (or if paranoid, change the else to else if (client->queue_last == req) {}

Reply:
Cancelling a queued request (which requires such lookup) should be a fairly rare operation, I decided not to optimize it. Most often, requests are being added to the tail of the queue (that's why there is `queue_last`) and removed one by one starting from the head of the queue (skipping the blocked ones). For the most typical use case the current data model is optimal, I think. If I smell a bottleneck, I'll consider re-doing it different way.

Reply:
Yes, ignore this, somehow I got it stuck in my head that the list was double linked.